File: CustomCollectionLayout/YMScrollerMenu/YWScrollerView.m
//
//  YWScrollerView.m
//  CustomCollectionLayout
//
//  Created by 天轩 on 2017/8/9.
//  Copyright © 2017年 天轩. All rights reserved.
//
#import "YWScrollerView.h"
/************************YWFlowLayout**************************/
@class YWFlowLayout;
@protocol YWFlowLayoutDelegate <NSObject>

@required
/**
 行数
 */
-(NSInteger)numberOfRowsInYWLayout:(YWFlowLayout *)YWLayout;
/**
 设置列数
 @parm 当数据源个数低于计算数目时
 此方法自动失效
 会以最合适的高度设置
 */
-(NSInteger)numberOfColumsInYWLayout:(YWFlowLayout *)YWLayout;

@end

@interface YWFlowLayout : UICollectionViewFlowLayout
@property(nonatomic,weak)id <YWFlowLayoutDelegate>delegate;
@property(nonatomic,readonly)NSInteger row;
@property(nonatomic,readonly)NSInteger colum;
@end

@implementation YWFlowLayout

-(void)prepareLayout
{
    //    准备方法被自动调用，以保证layout实例的正确。
    [super prepareLayout];
    self.scrollDirection=UICollectionViewScrollDirectionVertical;
    self.collectionView.showsVerticalScrollIndicator=NO;
    self.collectionView.showsHorizontalScrollIndicator=NO;
    self.collectionView.bounces=NO;
    self.collectionView.pagingEnabled=YES;
    
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfColumsInYWLayout:)])
    {
        _colum=[self.delegate numberOfColumsInYWLayout:self];
    }
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfRowsInYWLayout:)])
    {
        _row=[self.delegate numberOfRowsInYWLayout:self];
    }
    NSAssert(_row!=0&&_colum!=0, @"这样还有什么意思呢？行数和列数不能为0");
    /*
     当设置数量不正确时，我们自动给取合适的高度
     */
    NSInteger itemCount = [self.collectionView numberOfItemsInSection:0];
    if (itemCount<(_row*_colum))
    {
        _row=(itemCount%_colum==0?itemCount/_colum:itemCount/_colum+1);
    }
    
}
-(NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
{
    //.返回rect中的所有的元素的布局属性
    NSInteger numCount = [self.collectionView numberOfItemsInSection:0];
    
    NSMutableArray *array=[NSMutableArray new];
    for (int i =0; i<numCount; i++)
    {
        UICollectionViewLayoutAttributes *att=[self layoutAttributesForItemAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0]];
        [array addObject:att];
    }
    return array;
}
-(UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewLayoutAttributes *attr=[UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
    
    NSInteger rows = _row;
    NSInteger colum = _colum;
    NSInteger page = indexPath.item/(rows * colum);
    /*
     计算间距
     */
    //    CGFloat cellW =(self.collectionView.bounds.size.width -self.sectionInset.left - self.sectionInset.right - (colum - 1) * self.minimumInteritemSpacing)/colum;
    /*
     不计算间距
     */
    CGFloat cellW=self.collectionView.bounds.size.width/colum;
    
    NSInteger  itemRow = (indexPath.item - page*(colum*rows))/ colum;//行数
    NSInteger itemCol = (indexPath.item - page*(colum*rows))% colum;//列数
    CGFloat cellX = page * self.collectionView.bounds.size.width + itemCol * (cellW) + self.sectionInset.left;
    /*
     同上
     */
    //    CGFloat cellH = (self.collectionView.bounds.size.height - self.sectionInset.top - self.sectionInset.bottom - (rows - 1) * self.minimumLineSpacing) / rows;
    CGFloat cellH =self.collectionView.bounds.size.height/rows;
    CGFloat cellY = itemRow * (cellH + self.sectionInset.top) + self.sectionInset.top;
    attr.frame =CGRectMake(cellX, cellY, cellW, cellH);
    //    返回对应于indexPath的位置的cell的布局属性
    return attr;
}
-(CGSize)collectionViewContentSize
{
    NSInteger itemCount = [self.collectionView numberOfItemsInSection:0];
    NSInteger page =itemCount%(_colum*_row)==0? itemCount/(_colum*_row):
    itemCount/(_colum*_row)+1;
    return CGSizeMake(self.collectionView.bounds.size.width*page, self.collectionView.frame.size.height);
}
@end

/************************YWScrollerView**************************/
@interface YWScrollerView()<YWFlowLayoutDelegate>
@property(nonatomic,strong)Class RegisterCell;
@property(nonatomic,strong)UICollectionView *YWCollection;
@property(nonatomic,assign)BOOL isXib;
@property(nonatomic,readonly,assign)NSInteger colum;
@property(nonatomic,readonly,assign)NSInteger rows;
@property(nonatomic,assign)CGRect react;
@end
@implementation YWScrollerView
-(instancetype)initWithFrame:(CGRect)frame WithRegisterClass:(nullable Class)cell isXib:(BOOL)isXib
{
    if (self=[super initWithFrame:frame])
    {
        self.RegisterCell=cell;
        self.isXib=isXib;
        _react=frame;
        [self creatUI:frame];
        
    }
    return self;
}
-(void)creatUI:(CGRect)frame
{
    self.userInteractionEnabled=YES;
    YWFlowLayout *Layout=[[YWFlowLayout alloc]init];
    Layout.delegate=self;
    _YWCollection=[[UICollectionView alloc]initWithFrame:CGRectMake(0, 0, frame.size.width, frame.size.height) collectionViewLayout:Layout];
    _YWCollection.backgroundColor=[UIColor grayColor];
    _YWCollection.delegate=self;
    _YWCollection.dataSource=self;
    [self addSubview:_YWCollection];
    if (_isXib==YES)
    {
        [_YWCollection registerNib:[UINib nibWithNibName:NSStringFromClass(_RegisterCell) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass(self.RegisterCell)];
    }else
    {
     [_YWCollection registerClass:_RegisterCell forCellWithReuseIdentifier:NSStringFromClass(_RegisterCell)];
    }
}
#pragma mark-UICollection-datasoure
-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}
-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    if (self.dataSource)
    {
        return self.dataSource.count;
    }
    return 0;
}
-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell=[collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass(_RegisterCell) forIndexPath:indexPath];
    if (self.delegate&&[self.delegate respondsToSelector:@selector(RegisterClassCell:cellForItemAtIndexPath:)]) {
          cell= [self.delegate RegisterClassCell:cell cellForItemAtIndexPath:indexPath];
        }
    return cell;
}
-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.delegate&&[self.delegate respondsToSelector:@selector(YWScrollView:didSelectItemForIndex:)]) {
        [self.delegate YWScrollView:self didSelectItemForIndex:indexPath];
    }
    
}
#pragma mark-FlowLayout delegate
-(NSInteger)numberOfColumsInYWLayout:(YWFlowLayout *)YWLayout
{
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfColumsInYWscrollerView)])
    {
        
         _colum= [self.delegate numberOfColumsInYWscrollerView];
        return _colum;
    }
    return _colum;
}
-(NSInteger)numberOfRowsInYWLayout:(YWFlowLayout *)YWLayout
{
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfRowsInYWScrollerView)])
    {
        _rows=[self.delegate numberOfRowsInYWScrollerView];
        
        return _rows;
    }
    return _rows;
}
-(void)reloadData
{
    [_YWCollection reloadData];
}
-(void)updateFrame:(CGRect)frame
{
    _YWCollection.frame=frame;
    [self reloadData];
}
-(void)setIsShowPage:(BOOL)isShowPage
{
    _isShowPage=isShowPage;
    _YWCollection.frame=CGRectMake(0, 0, _react.size.width, _react.size.height-20);
    if (!_pageControl)
    {
        [self creatPageController];
    }
}
-(void)creatPageController
{
    NSInteger itemCount=[_YWCollection numberOfItemsInSection:0];
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfRowsInYWScrollerView)])
    {
        _rows=[self.delegate numberOfRowsInYWScrollerView];
    }
    if (self.delegate&&[self.delegate respondsToSelector:@selector(numberOfColumsInYWscrollerView)])
    {
        _colum= [self.delegate numberOfColumsInYWscrollerView];
    }
    NSInteger page =itemCount%(_colum*_rows)==0? itemCount/(_colum*_rows):
    itemCount/(_colum*_rows)+1;
    _pageControl= [[UIPageControl alloc] init];
    _pageControl.frame=CGRectMake([UIScreen mainScreen].bounds.size.width/2-120/2, CGRectGetHeight(_YWCollection.frame), 120, 20);
    _pageControl.backgroundColor=[UIColor colorWithRed:0 green:0 blue:0 alpha:0.5];
    _pageControl.numberOfPages =page;
    _pageControl.currentPage = 0;
    _pageControl.pageIndicatorTintColor = [UIColor clearColor];
    _pageControl.currentPageIndicatorTintColor = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.5];
    [self addSubview:_pageControl];
}
-(void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    _pageControl.currentPage=scrollView.contentOffset.x/CGRectGetWidth(_YWCollection.frame);
}
@end

File: Swift_Version/YWScrollerView.swift
//
//  YWScrollerView.swift
//  CustomCollectionLayout
//
//  Created by 天轩 on 2018/7/18.
//  Copyright © 2018年 天轩. All rights reserved.
//

import UIKit
let YWScrollIdent:String = "ywscrollIdent";
class YWScrollerView: UIView ,UICollectionViewDelegate,UICollectionViewDataSource{
    var collectionV:UICollectionView!
    var delegate:YWScrollerLayoutDelegate?
    var dataSource:[String]? {
        willSet{
//            print(newValue!);
        }
        didSet{
//            print(oldValue ?? []);
            //刷新数据
            self.collectionV.reloadData();
        }
    }
    init(frame: CGRect,registerCell:AnyClass,isXib:Bool,delegate:YWScrollerLayoutDelegate) {
        super.init(frame: frame);
        self.creatUI(frame: frame,registerCell
            :registerCell,isXib: isXib,delgate: delegate);
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    func creatUI(frame:CGRect,registerCell:AnyClass,isXib:Bool,delgate:YWScrollerLayoutDelegate)  {
        
        let flayLayout = YWScrollerLayout();
        self.delegate = delgate;
        if self.delegate != nil {
           flayLayout.rows = delegate!.numberOfRowsInYWLayout(layout: flayLayout);
           flayLayout.colum = delegate!.numberOfColumsInYWLayout(layout: flayLayout)
        }
        let collectonView = UICollectionView(frame: CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height), collectionViewLayout: flayLayout)
        collectonView.backgroundColor = UIColor.white;
        collectonView.delegate = self;
        collectonView.dataSource = self;
        self.addSubview(collectonView);
        if isXib == true {
            collectonView.register(UINib.init(nibName: NSStringFromClass(registerCell), bundle: .main), forCellWithReuseIdentifier: YWScrollIdent)
        }else{
             collectonView.register(registerCell, forCellWithReuseIdentifier: YWScrollIdent)
        }
       
        self.collectionV = collectonView;
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.dataSource?.count ?? 0;
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        var cell:UICollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: YWScrollIdent, for: indexPath);
        if self.delegate != nil {
            cell = (self.delegate?.YWScrollerCellForRow(registerCell: cell, IndexPath: indexPath))!;
        }
        return cell;
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        if self.delegate != nil {
            delegate?.YWScrollerViewDidSelect(index: indexPath);
        }
    }
}
protocol YWScrollerLayoutDelegate {
    func YWScrollerCellForRow(registerCell:UICollectionViewCell,IndexPath:IndexPath) -> UICollectionViewCell;
    //传入列数
    func numberOfRowsInYWLayout(layout:YWScrollerLayout) -> Int;
    //传入行数
    func numberOfColumsInYWLayout(layout:YWScrollerLayout) -> Int;
    //点击事件
    func YWScrollerViewDidSelect(index:IndexPath) -> Void;
}
class YWScrollerLayout: UICollectionViewFlowLayout {
    var rows:Int = 0;
    var colum:Int = 0;
    override func prepare() {
        super.prepare()
        self.scrollDirection = UICollectionViewScrollDirection.vertical;
        self.collectionView?.showsVerticalScrollIndicator = false;
        self.collectionView?.showsHorizontalScrollIndicator = false;
        self.collectionView?.bounces = false;
        self.collectionView?.isPagingEnabled = true;
        
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        //.返回rect中的所有的元素的布局属性
        let numCount = self.collectionView?.numberOfItems(inSection: 0) ?? 0;
        
        var array:[UICollectionViewLayoutAttributes] = [];
        for i in 0..<numCount {
            let att = self.layoutAttributesForItem(at: NSIndexPath(row: i, section: 0) as IndexPath)
            array.append(att!);
        }
        return array;
    }
    
    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        let attr:UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes.init(forCellWith: indexPath);
        let rows = self.rows;
        let colum = self.colum;
        let page = indexPath.item/(rows*colum)
        let cellW = (self.collectionView?.bounds.size.width)!/CGFloat(colum)
        let itemRow = (indexPath.item - page*(colum*rows))/colum
        let itemCol = (indexPath.item - page*(colum*rows))%colum;
        let cellX = CGFloat(page) * (self.collectionView?.bounds.size.width)!+CGFloat(itemCol)*cellW + self.sectionInset.left;
        
        /*不计算间距
         */
        //        CGFloat cellW=self.collectionView.bounds.size.width/colum;
        
        //        NSInteger  itemRow = (indexPath.item - page*(colum*rows))/ colum;//行数
        //        NSInteger itemCol = (indexPath.item - page*(colum*rows))% colum;//列数
        //        CGFloat cellX = page * self.collectionView.bounds.size.width + itemCol * (cellW) + self.sectionInset.left;
        /*
         同上
         */
        //    CGFloat cellH = (self.collectionView.bounds.size.height - self.sectionInset.top - self.sectionInset.bottom - (rows - 1) * self.minimumLineSpacing) / rows;
        //        CGFloat cellH =self.collectionView.bounds.size.height/rows;
        let cellH = (self.collectionView?.bounds.size.height)!/CGFloat(rows);
        //        CGFloat cellY = itemRow * (cellH + self.sectionInset.top) + self.sectionInset.top;
        let cellY = CGFloat(itemRow) * (cellH+self.sectionInset.top)+self.sectionInset.top;
        attr.frame = CGRect(x: cellX, y: cellY, width: cellW, height: cellH);
        
        return attr;
    }
    
    override var collectionViewContentSize: CGSize{
        let  itemCount = self.collectionView?.numberOfItems(inSection: 0);
        let page:Int;
        if itemCount!%(self.colum*self.rows) == 0 {
            page = itemCount!/(self.colum*self.rows);
        }else{
            page = itemCount!/(self.colum*self.rows)+1
        }
        return CGSize(width: self.collectionView!.bounds.size.width*CGFloat(page), height: self.collectionView!.frame.size.height);
    }
}
